Build-Wheels.py

```diff
@@ -0,0 +1,66 @@
+import json
+
+import requests
+from packaging.requirements import InvalidRequirement
+from packaging.requirements import Requirement
+
+# Global variables for branches and constraints files for the branches to be downloaded
+# (length of each list should be the same and so the positions for branch and it's specific constraints file)
+IDF_BRANCH = ['release/v5.0', 'release/v5.1', 'master']
+IDF_CONSTRAINTS = ['v5.0', 'v5.1', 'v5.2']
+
+# TODO temporary solution
+install_optional = True
+
+
+idf_resources_url = 'https://raw.githubusercontent.com/espressif/esp-idf/'
+requirements_txt = []
+requirements_set = set()
+
+# Download all requirements from all the branches requirements and constraints files
+for i in range(len(IDF_BRANCH)):
+    idf_requirements_json_url = f'{idf_resources_url}{IDF_BRANCH[i]}/tools/requirements.json'
+    constraint_file_url = f'https://dl.espressif.com/dl/esp-idf/espidf.constraints.{IDF_CONSTRAINTS[i]}.txt'
+
+    r = requests.get(idf_requirements_json_url)
+    if r.status_code == 200:
+        idf_requirements_json = json.loads(r.content)
+    else:
+        print('Failed to download requirements JSON')
+
+    # Download requirements files from all requirements groups specified in JSON
+    print(f'---------- ESP-IDF BRANCH {IDF_BRANCH[i]} ----------')
+    for feature in idf_requirements_json['features']:
+        r = requests.get(f"{idf_resources_url}{IDF_BRANCH[0]}/{feature['requirement_path']}")
+        if r.status_code == 200:
+            if not feature['optional'] or install_optional:
+                requirements_txt += r.text.splitlines()
+                print(f"Added ESP-IDF {feature['name']} requirements")
+        else:
+            print(f"Failed to download feature (requirement group) '{feature['name']}'")
+
+    # Download constraints file for specific branch
+    r = requests.get(constraint_file_url)
+    if r.status_code == 200:
+        requirements_txt += r.text.splitlines()
+        print(f'Added ESP-IDF constraints file for branch {IDF_BRANCH[i]}')
+    else:
+        print(f'Failed to download ESP-IDF constraints for branch {IDF_BRANCH[i]}')
+
+    # Add requirements into set to prevent duplicates
+    for line in requirements_txt:
+        if '#' not in line and line != '\n':
+            try:
+                requirements_set.add(Requirement(line.strip()))
+            except InvalidRequirement:
+                # TODO not a classic requirement
```

Comment:
I am just curious, what is example of 'not a classic requirement'? 🤔 Maybe it would be nice to include a comment with example when you finish the implementation

Reply:
Yes, that is a good idea. I am not sure how this will be implemented yet.

The example of 'not a classic requirement' is for example: `--only-binary cryptography`